### PR TITLE
[Routing][FrameworkBundle] Allow configurable query encoding type in UrlGenerator

### PR DESCRIPTION
**Pull request by likeuntomurphy**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        | symfony/symfony-docs#10073

This PR introduces a change that makes it very easy for developers to use RFC 1738 if they choose.

I have seen #19625, but I believe there are cases when PHP_QUERY_RFC1738 may be preferred even if the generator is not used in a form context, strictly speaking. One such case is search — it can begin with an actual form, but subsequent facet and pagination links can simply be generated. Pagination and related search links in Google use RFC 1738, as do pagination and “Pro Tip!” recommendation links on GitHub.

While we could simply replace `%20` with `+` after generating facet URLs, it becomes much more complicated to accomplish the same effect in pagination links if you are using a bundle like KnpPaginatorBundle.

Extending the generator class seems excessive, as it would require copying over 150 lines of code to override the `doGenerate` method just to change one parameter.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php
Comment (iltar) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: This info is only correct for the default value, it should probably be something more generic.

How many possibilities are there for this option? Because it would be really nice if Symfony would tell you when using an invalid value. Perhaps you can limit it to the possible methods using validation?

File: src/Symfony/Component/Routing/Generator/ConfigurableQueryEncTypeInterface.php
Comment (iltar) [inline src/Symfony/Component/Routing/Generator/ConfigurableQueryEncTypeInterface.php]: This method is not used, should it really be added?

File: src/Symfony/Component/Routing/Generator/ConfigurableQueryEncTypeInterface.php
Comment (iltar) [inline src/Symfony/Component/Routing/Generator/ConfigurableQueryEncTypeInterface.php]: For clarity I recommend to name it `ConfigurableQueryEncodingTypeInterface` with the method `setQueryEncodingType`, I'm not sure if `EncType` is a standard name here though.

**Review by linaori — COMMENTED** [review]
I'm not entirely sure if this covers all use-cases. What if you generate a URL to an external page which doesn't follow the same encoding type? Should we add a reserved key for that in the parameters list perhaps?

**Review by likeuntomurphy — COMMENTED** [review]
Thanks for the feedback!

I understand your concern, but I am not convinced it falls within the scope of this change. Because the URL generator currently uses a hard-coded value, not being able to change the encoding type is already a limitation. I am less concerned with that limitation than I am with being able to change the encoding type deep inside the URL generator class in a simple way.

That said, you can actually now change the encoding type on the fly if you are so inclined. For instance, it is now possible to do something like, 

```php
$originalEncoding = $generator->getQueryEncodingType();
$generator->setQueryEncodingType($preferredEncoding);
$url = $generator->generateUrl(...$args);
$generator->setQueryEncodingType($originalEncoding);
```

I am not convinced that is the most elegant way of handling this situation, but it is the reason the `getQueryEncodingType` method is included in the `ConfigurableQueryEncodingTypeInterface`.

File: src/Symfony/Component/Routing/Generator/ConfigurableQueryEncodingTypeInterface.php
Comment (nicolas-grekas) [inline src/Symfony/Component/Routing/Generator/ConfigurableQueryEncodingTypeInterface.php]: I'm not sure adding this interface is a good idea. What about adding a constructor argument instead?

**Review by nicolas-grekas — COMMENTED** [review]
Can you decorate the generator and simply `str_replace('%20', '+', parent::doGenerate(...)` to achieve the same?

**Comment by likeuntomurphy** [comment]
It would require a more complex block of code than that in order to execute the replacement only on the query component of the URL — `str_replace` would also catch `%20` in the path and fragment components, which is not desired.

So it could be something like this:

```php
$url = parent::doGenerate(...);

$parts = parse_url($url);

if (isset($parts['query'])) {
    $search = $parts['query'];
    $replace = str_replace('%20', '+', $search);
    $url = str_replace($search, $replace, $url);
}
```
For reliability I would be more comfortable with something like

```php
if (isset($parts['query'])) {
    $parts['query'] = str_replace('%20', '+', $parts['query']);

    $url = '';
    
    if (isset($parts['scheme'])) {
        $url .= $parts['scheme'];
    }
    
    if (isset($parts['host'])) {
        ...
    }
    
    if (isset($parts['port'])) {
        ...
    }
    
    ...
}
```
But to me that starts to be excessively redundant with the heavy lifting `doGenerate` has already done, and it seems much preferable to simply have a way to toggle the parameter passed to `http_build_query`.

File: src/Symfony/Component/Routing/Generator/ConfigurableQueryEncodingTypeInterface.php
Comment (likeuntomurphy) [reply src/Symfony/Component/Routing/Generator/ConfigurableQueryEncodingTypeInterface.php]: My understanding was that adding an argument would be a BC break for anyone who may have extended the `UrlGenerator`, so I mirrored the way configurable requirements were implemented. I see that `defaultEncoding` was added in `4.1` (it wasn’t there when I wrote this in December) — so I’ll happily change it if my understanding of the BC implications is wrong.

Comment (nicolas-grekas) [reply src/Symfony/Component/Routing/Generator/ConfigurableQueryEncodingTypeInterface.php]: adding argument *on the constructor* is fone

Comment (likeuntomurphy) [reply src/Symfony/Component/Routing/Generator/ConfigurableQueryEncodingTypeInterface.php]: And after more than a moment’s thought, the reason behind that is perfectly clear. Thanks!

**Comment by Tobion** [comment]
I don't think a global config is appropriate for this. Using RFC 1738 depends on the mime type of the page and you could even mix links with encoding RFC 1738 and RFC 3986 in the same page if you need to. So a global config would not fit there.
For this reason, it seems better to just write a helper function that replaces %20 with + as you pointed out and does not require a change in route generation.

> While we could simply replace %20 with + after generating facet URLs, it becomes much more complicated to accomplish the same effect in pagination links if you are using a bundle like KnpPaginatorBundle.

You can provide a custom template for the pagination in the KnpPaginatorBundle that uses + for spaces: https://github.com/KnpLabs/KnpPaginatorBundle/blob/f4ece5b347121b9fe13166264f197f90252d4bd2/Twig/Extension/PaginationExtension.php#L46
Or you could write a response filter that just replaces all links in the pagination div.

**Comment by Tobion** [comment]
Btw, the PR with the implementation and description is excellent. Good job! I just don't think this is the right solution to the problem.

**Comment by likeuntomurphy** [comment]
Thanks, @tobion.

I was trying to make a good argument by anticipating counterarguments — but ultimately, I would not have created the pull request if I did not feel that the framework should let the developer have control over which query encoding type is used and avoid complex workarounds. I understand that RFC 1738 may be considered outdated, and RFC 3986 is the default for good reason. But I want to copy what Google does! 😄 

    you could even mix links with encoding RFC 1738 and RFC 3986 in the same page
    if you need to.

This is a point @iltar also raised, and it’s a good one. But I don’t understand how it is a relevant criticism of this PR, because you can’t do this currently. 

As I see it, there are four options:

1) as of #19625 — the generator always uses RFC 1738
2) as of #19639 — the generator always uses RFC 3986
3) this PR — the generator always uses the developer’s preferred encoding type
4) TBD — the generator lets you specify the encoding type at runtime (as a method arg perhaps?)

Are you suggesting we skip 3 and proceed straight to 4? Or do you feel that the framework should continue to use RFC 3986 exclusively and leave it to developers to find alternate ways of achieving RFC 1738 links?

**Comment by nicolas-grekas** [comment]
Honestly, I'm fine with this feature. I don't see why we should force RFC 3986 over RFC 1738. Changing globally is fine to me. At least there are no practical downsides I know about. @Tobion I feel like you're suggesting there are differences related to the mime content type, can you elaborate?

**Comment by nicolas-grekas** [comment]
Can't we make RFC1738 the hardcoded behavior instead of RFC 3986 actually, for the query part only?

**Comment by likeuntomurphy** [comment]
That was the original behavior before #19639 was merged, since the `enc_type` in `http_build_query` defaults to `PHP_QUERY_RFC1738`.

Undoing that change would nullify the discussion and decisions in #19625 and #19639. It would also create a BC break for anyone who made adjustments because of the previous BC break (#20856).

**Review by nicolas-grekas — APPROVED** [review]
Time to move forward here.
OK on my side. @symfony/deciders?

**Comment by Tobion** [comment]
Global configs are a dead end IMO. Let's say I write a reusable bundle and generate links to a controller. I cannot even assume anymore that I can read read links properly that I generated myself.

**Comment by lyrixx** [comment]
I agree with @Tobion here

**Pull request closed without merging**
